File: Serene/Serene.Script/Common/Navigation/LanguageSelection.cs
namespace Serene.Common
{
    using Serene.Administration;
    using jQueryApi;
    using Serenity;
    using System.Html;
    using System.Linq;

    public class LanguageSelection : Widget
    {
        public LanguageSelection(jQueryObject select, string currentLanguage)
            : base(select)
        {
            currentLanguage = currentLanguage ?? "en";

            var self = this;
            this.Change(e =>
            {
                jQuery.Instance.cookie("LanguagePreference", select.GetValue(), new {
                    path = Q.Config.ApplicationPath,
                    expires = 365
                });
                Window.Location.Reload(true);
            });

            Q.GetLookupAsync<LanguageRow>("Administration.Language").Then(x =>
            {
                if (!x.Items.Any(z => z.LanguageId == currentLanguage))
                {
                    var idx = currentLanguage.LastIndexOf("-");
                    if (idx >= 0)
                    {
                        currentLanguage = currentLanguage.Substr(0, idx);
                        if (!x.Items.Any(z => z.LanguageId == currentLanguage))
                        {
                            currentLanguage = "en";
                        }
                    }
                    else
                        currentLanguage = "en";
                }

                foreach (var l in x.Items)
                    Q.AddOption(select, l.LanguageId, l.LanguageName);

                select.Value(currentLanguage);
            });
        }
    }
}

File: Serene/Serene.Core/Modules/Common/Navigation/LanguageSelection.ts
namespace Serene.Common {
    export class LanguageSelection extends Serenity.Widget<any> {
        constructor(select: JQuery, currentLanguage: string) {
            super(select);

            currentLanguage = Q.coalesce(currentLanguage, 'en');

            this.change(e => {
                var path = Q.Config.applicationPath;
                if (path && path != '/' && Q.endsWith(path, '/'))
                    path = path.substr(0, path.length - 1);
                $.cookie('LanguagePreference', select.val(), {
                    path: path,
                    expires: 365
                });
                window.location.reload(true);
            });

            Q.getLookupAsync<Administration.LanguageRow>('Administration.Language').then(x => {
                if (!Q.any(x.items, z => z.LanguageId === currentLanguage)) {
                    var idx = currentLanguage.lastIndexOf('-');
                    if (idx >= 0) {
                        currentLanguage = currentLanguage.substr(0, idx);
                        if (!Q.any(x.items, y => y.LanguageId === currentLanguage)) {
                            currentLanguage = 'en';
                        }
                    }
                    else {
                        currentLanguage = 'en';
                    }
                }

                for (var l of x.items) {
                    Q.addOption(select, l.LanguageId, l.LanguageName);
                }

                select.val(currentLanguage);
            });
        }
    }
}